File: packages/adl-agent/implementations/google-adk-ts/agent.ts
/**
 * ADL Explainer Agent — Google Generative AI SDK (TypeScript) implementation.
 *
 * Uses the `@google/generative-ai` package for the Gemini tool-use loop.
 * Tools and knowledge are shared from the parent src/ directory.
 */

import type {
  ChatSession,
  FunctionDeclaration,
  FunctionDeclarationSchema,
  GenerateContentResult,
  GenerativeModel,
  Part,
  Tool,
} from "@google/generative-ai";
import { SchemaType } from "@google/generative-ai";
import { SYSTEM_PROMPT, TOOLS, executeTool, type ToolDefinition } from "../../src/shared.js";

// Re-export shared items for consumers of this implementation
export { SYSTEM_PROMPT, TOOLS, executeTool };

/**
 * Map a JSON Schema type string to Gemini SchemaType enum.
 */
function toGeminiType(jsonType: string): SchemaType {
  switch (jsonType) {
    case "string":
      return SchemaType.STRING;
    case "number":
      return SchemaType.NUMBER;
    case "integer":
      return SchemaType.INTEGER;
    case "boolean":
      return SchemaType.BOOLEAN;
    case "array":
      return SchemaType.ARRAY;
    case "object":
      return SchemaType.OBJECT;
    default:
      return SchemaType.STRING;
  }
}

/**
 * Convert a shared ToolDefinition's parameter properties to Gemini schema format.
 */
function toGeminiParameters(tool: ToolDefinition): FunctionDeclarationSchema {
  const properties: Record<string, { type: SchemaType; description?: string; enum?: string[] }> = {};

  for (const [key, value] of Object.entries(tool.parameters.properties)) {
    const prop = value as { type: string; description?: string; enum?: string[] };
    const geminiProp: { type: SchemaType; description?: string; enum?: string[] } = {
      type: toGeminiType(prop.type),
    };
    if (prop.description) {
      geminiProp.description = prop.description;
    }
    if (prop.enum) {
      geminiProp.enum = prop.enum;
    }
    properties[key] = geminiProp;
  }

  return {
    type: SchemaType.OBJECT,
    properties,
    required: tool.parameters.required,
  } as FunctionDeclarationSchema;
}

/**
 * Map shared ToolDefinitions to Gemini tool format.
 */
function toGeminiTools(tools: ToolDefinition[]): Tool[] {
  const functionDeclarations: FunctionDeclaration[] = tools.map((t) => ({
    name: t.name,
    description: t.description,
    parameters: toGeminiParameters(t),
  }));

  return [{ functionDeclarations }];
}

export const GEMINI_TOOLS = toGeminiTools(TOOLS);

export interface AgentConfig {
  maxOutputTokens?: number;
}

/**
 * Check if a GenerateContentResult has function call parts.
 */
function getFunctionCalls(result: GenerateContentResult): Array<{ name: string; args: Record<string, unknown> }> {
  const calls: Array<{ name: string; args: Record<string, unknown> }> = [];

  for (const candidate of result.response.candidates ?? []) {
    for (const part of candidate.content.parts) {
      if (part.functionCall) {
        calls.push({
          name: part.functionCall.name,
          args: (part.functionCall.args ?? {}) as Record<string, unknown>,
        });
      }
    }
  }

  return calls;
}

/**
 * Extract final text from a GenerateContentResult.
 */
function getTextContent(result: GenerateContentResult): string {
  const texts: string[] = [];

  for (const candidate of result.response.candidates ?? []) {
    for (const part of candidate.content.parts) {
      if (part.text) {
        texts.push(part.text);
      }
    }
  }

  return texts.join("\n");
}

/**
 * Run a single turn of the agent — send a message, handle tool calls, return the final text.
 */
export async function runAgentTurn(
  model: GenerativeModel,
  chat: ChatSession,
  userMessage: string,
  config?: AgentConfig,
): Promise<string> {
  let result = await chat.sendMessage(userMessage);
  let functionCalls = getFunctionCalls(result);

  while (functionCalls.length > 0) {
    // Execute each function call and build response parts
    const responseParts: Part[] = functionCalls.map((call) => {
      const toolResult = executeTool(call.name, call.args);
      return {
        functionResponse: {
          name: call.name,
          response: { result: JSON.parse(toolResult) },
        },
      };
    });

    // Send function responses back
    result = await chat.sendMessage(responseParts);
    functionCalls = getFunctionCalls(result);
  }

  return getTextContent(result);
}

File: packages/adl-agent/implementations/google-adk-py/agent.py
"""
ADL Explainer Agent -- Google Generative AI Python SDK implementation.

Uses the google-genai package for the Gemini tool-use loop.
"""

from __future__ import annotations

import json
from typing import Any

from google import genai
from google.genai import types

from tools import execute_tool

SYSTEM_PROMPT = (
    "You are the ADL Spec Explainer, an expert on the Agent Definition Language (ADL).\n"
    "\n"
    "You help users:\n"
    "- Understand ADL concepts (the passport model, tools, permissions, data classification, "
    "profiles, lifecycle, security, extensions)\n"
    "- Validate their ADL documents and explain any errors in plain language\n"
    "- Explore example ADL documents\n"
    "- Navigate the ADL specification\n"
    "- Compare ADL with other agent description formats (A2A, MCP, OpenAPI, Agent Protocol)\n"
    "\n"
    "Always be clear, accurate, and helpful. When explaining validation errors, provide "
    "actionable guidance on how to fix them. Use the available tools to look up information "
    "rather than relying on memory alone.\n"
    "\n"
    "ADL documents are YAML or JSON files that describe AI agents -- like a passport that "
    "travels with the agent, declaring its identity, capabilities, permissions, and trust signals."
)

# Build Gemini function declarations from tool definitions
_TOOL_DECLARATIONS: list[types.FunctionDeclaration] = [
    types.FunctionDeclaration(
        name="explain_concept",
        description=(
            "Explain an ADL concept such as the passport model, tools, permissions, "
            "data classification, profiles, lifecycle, security, or extensions."
        ),
        parameters=types.Schema(
            type=types.Type.OBJECT,
            properties={
                "concept": types.Schema(
                    type=types.Type.STRING,
                    description="The ADL concept to explain",
                ),
            },
            required=["concept"],
        ),
    ),
    types.FunctionDeclaration(
        name="validate_document",
        description=(
            "Validate an ADL document provided as YAML or JSON and return structured "
            "results with errors explained in plain language."
        ),
        parameters=types.Schema(
            type=types.Type.OBJECT,
            properties={
                "document": types.Schema(
                    type=types.Type.STRING,
                    description="The ADL document content as YAML or JSON string",
                ),
            },
            required=["document"],
        ),
    ),
    types.FunctionDeclaration(
        name="show_example",
        description="Return an example ADL document by category: minimal, production, or with-tools.",
        parameters=types.Schema(
            type=types.Type.OBJECT,
            properties={
                "category": types.Schema(
                    type=types.Type.STRING,
                    description="The example category to retrieve",
                    enum=["minimal", "production", "with-tools"],
                ),
            },
            required=["category"],
        ),
    ),
    types.FunctionDeclaration(
        name="get_spec_section",
        description="Retrieve information about a specific section of the ADL specification.",
        parameters=types.Schema(
            type=types.Type.OBJECT,
            properties={
                "section": types.Schema(
                    type=types.Type.STRING,
                    description="The spec section or topic to look up",
                ),
            },
            required=["section"],
        ),
    ),
    types.FunctionDeclaration(
        name="compare_formats",
        description=(
            "Compare ADL with another agent description format such as A2A, MCP, "
            "OpenAPI, or Agent Protocol."
        ),
        parameters=types.Schema(
            type=types.Type.OBJECT,
            properties={
                "format": types.Schema(
                    type=types.Type.STRING,
                    description="The format to compare against ADL",
                    enum=["a2a", "mcp", "openai", "agent_protocol"],
                ),
            },
            required=["format"],
        ),
    ),
]

TOOLS = types.Tool(function_declarations=_TOOL_DECLARATIONS)


def _get_function_calls(
    response: types.GenerateContentResponse,
) -> list[tuple[str, dict[str, Any]]]:
    """Extract function call parts from a Gemini response."""
    calls: list[tuple[str, dict[str, Any]]] = []
    for candidate in response.candidates or []:
        for part in candidate.content.parts or []:
            if part.function_call:
                args = dict(part.function_call.args) if part.function_call.args else {}
                calls.append((part.function_call.name, args))
    return calls


def _get_text_content(response: types.GenerateContentResponse) -> str:
    """Extract text content from a Gemini response."""
    texts: list[str] = []
    for candidate in response.candidates or []:
        for part in candidate.content.parts or []:
            if part.text:
                texts.append(part.text)
    return "\n".join(texts)


def run_agent_turn(
    client: genai.Client,
    chat: Any,
    user_message: str,
    model_name: str = "gemini-2.0-flash",
) -> str:
    """Run a single agent turn: send a message, handle tool calls, return final text."""
    response = chat.send_message(user_message)
    function_calls = _get_function_calls(response)

    while function_calls:
        # Execute each function call and build response parts
        function_responses = []
        for name, args in function_calls:
            result_str = execute_tool(name, args)
            result_data = json.loads(result_str)
            function_responses.append(
                types.Part.from_function_response(
                    name=name,
                    response={"result": result_data},
                )
            )

        # Send function responses back
        response = chat.send_message(function_responses)
        function_calls = _get_function_calls(response)

    return _get_text_content(response)
